src/main/java/org/broadinstitute/hellbender/utils/runtime/StreamingProcessController.java

```diff
@@ -230,15 +239,50 @@ public boolean waitForAck() {
 
         try {
             // blocking call to wait for the ack
-            boolean isAck = ackFuture.get();
-            processJournal.writeLogMessage(isAck ? ACK_LOG_MESSAGE : NCK_LOG_MESSAGE);
+            final ProcessControllerAckResult pcAck = ackFuture.get();
+            processJournal.writeLogMessage(pcAck.getDisplayMessage());
             ackFuture = null;
-            return isAck;
+            return pcAck;
         } catch (InterruptedException | ExecutionException e) {
             throw new GATKException("Exception waiting for ack from Python: " + e.getMessage(), e);
         }
     }
 
+    // Retrieve a nkm message from the input stream and package it up as a ProcessControllerAckResult
+    private ProcessControllerAckResult getNckWithMessageResult() throws IOException {
+        // look for a 4 byte long string with a 4 byte decimal integer with a value < 9999
+        final byte[] nckMessageLengthBuffer = new byte[NCK_WITH_MESSAGE_MESSAGE_LEN_SIZE];
+        final int nMessageLengthBytes = ackFIFOInputStream.read(nckMessageLengthBuffer, 0, NCK_WITH_MESSAGE_MESSAGE_LEN_SIZE);
```

Comment:
Yeah - thats a good catch. The whole reason I chose a constant (4) for this length was to make this process of consuming it deterministic, and the Python side flushes the output stream after it writes the length and message. But I think you're right that there is still a chance that `read` could return before having consumed the whole 4 bytes.  Luckily, I can hijack the `getNckMessage` method below, which was originally for retrieving the message string, and also use it to retrieve the length, since it already loops until it gets the number of bytes its expecting. Will need to rename it though.

Reply:
Done. I updated all of the places where we read from the stream now to use the (newly renamed) method `getBytesFromStream` to make sure we always either get the expected number of bytes, or throw.